core/src/storage2/lazy/lazy_hmap.rs

```diff
@@ -814,40 +862,77 @@ where
+            EntryOrMutableValue::MutableValue(v_mut) => {
+                core::mem::replace(v_mut.value_mut(), None)
+                    .expect("entry behind `MutableValue` must always exist")
+            }
+        };
         (self.key, old)
     }
 
     /// Gets a reference to the value in the entry.
     pub fn get(&self) -> &V {
-        self.entry
-            .get()
-            .value()
-            .as_ref()
-            .expect("entry behind `OccupiedEntry` must always exist")
+        match &self.entry {
+            EntryOrMutableValue::Entry(entry) => {
+                entry
+                    .get()
+                    .value()
+                    .as_ref()
+                    .expect("entry behind `OccupiedEntry` must always exist")
+            }
+            EntryOrMutableValue::MutableValue(v_mut) => {
+                v_mut
+                    .value()
+                    .as_ref()
+                    .expect("entry behind `MutableValue` must always exist")
+            }
+        }
     }
 
     /// Gets a mutable reference to the value in the entry.
     ///
     /// If you need a reference to the `OccupiedEntry` which may outlive the destruction of the
     /// `Entry` value, see `into_mut`.
     pub fn get_mut(&mut self) -> &mut V {
-        self.entry
-            .get_mut()
-            .value_mut()
-            .as_mut()
-            .expect("entry behind `OccupiedEntry` must always exist")
+        match &mut self.entry {
+            EntryOrMutableValue::Entry(entry) => {
+                entry
+                    .get_mut()
+                    .value_mut()
+                    .as_mut()
+                    .expect("entry behind `OccupiedEntry` must always exist")
+            }
+            EntryOrMutableValue::MutableValue(v_mut) => {
+                v_mut
+                    .value_mut()
+                    .as_mut()
+                    .expect("entry behind `MutableValue` must always exist")
+            }
+        }
     }
 
     /// Sets the value of the entry, and returns the entry's old value.
     pub fn insert(&mut self, new_value: V) -> V {
-        let new_value = Box::new(StorageEntry::new(Some(new_value), EntryState::Mutated));
-        self.entry
-            .insert(new_value)
-            .into_value()
-            .expect("entry behind `OccupiedEntry` must always exist")
+        match &mut self.entry {
+            EntryOrMutableValue::Entry(entry) => {
+                let new_value =
+                    Box::new(StorageEntry::new(Some(new_value), EntryState::Mutated));
+                entry
+                    .insert(new_value)
+                    .into_value()
+                    .expect("entry behind `OccupiedEntry` must always exist")
+            }
+            EntryOrMutableValue::MutableValue(v_mut) => {
+                core::mem::replace(v_mut.value_mut(), Some(new_value))
+                    .expect("entry behind `MutableValue` must always exist")
```

Comment:
I am not sure right now about this but this routine should make sure that the value is not only replaced but also that the cache is notified that it is no longer preserving the value if this is not already the case so that the cache properly flushes the value to the storage upon the end of execution.

Reply:
It's flushed, I added a test for this now here: https://github.com/paritytech/ink/pull/480/commits/f5d011df1b948902610d5d77012c03f565b67ae9.